src/impl/vamp/planning/rrt_star.hh

```diff
@@ -0,0 +1,285 @@
+            // main loop
+            std::size_t iter = 0;
+            std::vector<std::pair<NNNode<dimension>, float>> near;
+            while (iter++ < settings.max_iterations and free_index < settings.max_samples)
+            {
+                // sample random config
+                auto sample_config = rng.next();
+                Robot::scale_configuration(sample_config);
+                typename Robot::ConfigurationBuffer sample_config_arr;
+                sample_config.to_array(sample_config_arr.data());
+
+                const auto nearest = tree.nearest(NNFloatArray<dimension>{sample_config_arr.data()});
+                if (not nearest)
+                {
+                    continue;
+                }
+
+                // steer (calculate next state within range)
+                const auto &[nearest_node, nearest_distance] = *nearest;
+                auto nearest_configuration = nearest_node.as_vector();
+                auto nearest_vector = sample_config - nearest_configuration;
+
+                bool reach = nearest_distance < settings.range;
+                auto extension_vector =
+                    (reach) ? nearest_vector : nearest_vector * (settings.range / nearest_distance);
+
+                if (validate_vector<Robot, rake, resolution>(
+                        nearest_configuration,
+                        extension_vector,
+                        (reach) ? nearest_distance : settings.range,
+                        environment))
+                {
+                    // insert new config into vertex set
+                    float *new_configuration_ptr = buffer_index(free_index);
+                    auto new_configuration = nearest_configuration + extension_vector;
+                    new_configuration.to_array(new_configuration_ptr);
+                    NNNode<dimension> new_node({free_index, {new_configuration_ptr}});
+
+                    // sample points in graph within radius r of new config
+                    const std::size_t card =
+                        tree.size() + 1;  // plus 1 to account for the new configuration not inserted yet
+                    const float rrt_radius = std::fmin(
+                        settings.rewire_factor * gamma_rrt * pow((log(card) / card), dim_recip),
+                        settings.range);
+                    tree.nearest(near, NNFloatArray<dimension>{new_configuration_ptr}, card, rrt_radius);
+
+                    // initialize variables to keep track of the min cost neighbor and the cost of that
+                    // neighbor
+                    auto min_neighbor = nearest_node;
+                    float min_cost = cost[nearest_node.index] + nearest_distance;
+
+                    // loop through near neighbors, find the one that gives us min cost from root
+                    std::vector<bool> collision_free(near.size());
+                    std::size_t idx = 0;
+                    for (auto &[node, distance] : near)
```

Comment:
`const auto &`?

Reply:
Also, I would just iterate through with `idx` as a loop variable (auto i = 0U; ...), and access near with:
```
const auto &[node, distance] = near[idx];
```